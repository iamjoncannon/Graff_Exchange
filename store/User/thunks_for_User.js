import actions from "./actions_for_User"
import axios from 'axios'
import { urlPrefix } from '../../secrets'

export const loginThunk = (email, password) => async dispatch => {

  let res 

  try{
    
    res = await axios.post( (urlPrefix + '/login'), { email, password} )

  }
  catch(error){

      alert(error.response.data.message)
  }
  
  dispatch(actions.login(JSON.parse(res.data)))
  
};

export const registerThunk = (Name, Email, Password) => async dispatch => {
 
  console.log(Name, Email, Password)

  const signUpInfo = { Name, Email, Password }

  let response

  try {

    response = await axios.post(urlPrefix + '/signup', signUpInfo )

    console.log("register response: ", response)
  }
  catch(error){
    
    let  duplicateEmail = error.response.data.message === `pq: duplicate key value violates unique constraint "unique_email"`

    if (duplicateEmail){

      alert("This Email is already on file, please try again with a unique email.")
    }
    else{

      console.log(error)

    }
  }

  signUpInfo.password = null 

  let returnedToken = response.data

  dispatch(actions.login({Balance: 5000, token: returnedToken, Name, Email}))
};

export default {
	loginThunk,
	registerThunk,
}