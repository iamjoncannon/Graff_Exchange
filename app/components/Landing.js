import React from 'react';
import { blurb, logoUrl } from './utils'
import { connect } from "react-redux";
import { Redirect } from "react-router-dom";
import { loginThunk, registerThunk } from "../../store/User/thunks_for_User.js";


class Landing extends React.Component {
 
  constructor(props){

      super(props)
      
      this.state = {
          mode : 'sign-in',
          firstName: "Webster",
          lastName: "Ross",
          email : "webster@ross.com",
          password : "password",
          submitted: false 
      }
  }

  componentDidMount(){

    // the background needs to be cream for the desktop to seem inset
    // ...but only for this page

    document.body.style.backgroundColor = "#F9FAFB" 
  }

  componentWillUnmount(){

    document.body.style.backgroundColor = 'white' 
  }

  DemoAccount = () => {

    this.setState({
          mode : 'sign-in',
          email : "webster@ross.com",
          password : "password"
    })

    setTimeout(()=>this.handleSubmit(), 50)
  }

  handleChange = (evt) => {

      this.setState({
          [evt.target.name]:evt.target.value
      })
  }

  handleSubmit = () => {

      let { firstName, lastName, email, password } = this.state

      if(this.state.mode === "sign-in"){

        this.props.handleLogin(email, password)
      }

      if(this.state.mode === "sign-up"){
        
        this.props.handleRegister(firstName, lastName, email, password)
      }

      this.setState({submitted: true})
  }

  render() {

    const isDesktop = window.innerWidth > 1100;

    const { mode } = this.state
    
    if(this.props.isLoggedIn){
        
        return (
            <Redirect to="/all/watchlist" />
        )
    }
    else {

      return (

        <div className={ isDesktop && mode === "sign-up" ? "landing landing-rev" : "landing"} >
        
          <div className="header">

              <span className="title">
                Graff Exchange
              </span>

              <span className="blurb">

                {blurb}
                
              </span>
              
              <img src={logoUrl} />

              {!isDesktop &&

                <div className="sign-btns">

                  <span 
                    className={ mode === "sign-up" ? "unselected" : undefined}
                    onClick={()=>this.setState({mode: "sign-in"})}
                  >             
                      Sign In
                    </span>

                  <span 
                    className={ mode === "sign-in" && !isDesktop ? "unselected" : undefined}
                    onClick={()=>this.setState({mode: "sign-up"})}
                  >
                    Sign Up
                  </span>

                </div>
              }

              {isDesktop && mode === "sign-up" && 

                  <div className="sign-btns">
                    
                      <span > Already have an account? </span> 

                    
                      <span 
                        onClick={()=>this.setState({mode: "sign-in"})}
                      >             
                          Sign In
                      </span> 
                  </div>
            }

            { isDesktop && mode == "sign-in" && 

              <div className="sign-btns">

                <span>Don't have an account? </span> 

                <span 
                  onClick={()=>this.setState({mode: "sign-up"})}
                >
                  Sign Up
                </span>

              </div>
            }

          </div>

          <div className="form">

            {
              mode === "sign-up" &&

              <div>

                <span>First Name</span>

                <input 
                  type="text"
                  name="firstName"
                  value={this.state.firstName}
                  onChange={this.handleChange} 
                />

              </div>
            }
            
            {
              mode === "sign-up" &&

              <div>

                <span>Last Name</span>

                <input 
                  type="text"
                  name="lastName"
                  value={this.state.lastName}
                  onChange={this.handleChange} 
                />

              </div>         
            }

            <div>

              <span>Email</span>

              <input
                type="text"
                name="email"
                value={this.state.email}
                onChange={this.handleChange} 
              />

            </div>

            <div>

              <span>Password</span>

              <input 
                onChange={this.handleChange}
                type="password"
                name="password"
                value={this.state.password}
              />

            </div>

          {

            mode === "sign-up" && 

            <div>

              <span>Confirm Password</span>

              <input 
                
                type="password"
              />
            </div>
          }

          <button
            onClick={ this.handleSubmit }
          >
              {
                mode === 'sign-in' && !this.state.submitted ? "Sign In" : 
                mode === 'sign-in' && !this.state.submitted ? "Sign Up" :
                this.state.submitted ? "Loading..." : ""
              }

          </button>

          { mode === "sign-in" && <span onClick={this.DemoAccount}>Sign into Demo Account</span> }
          
          </div>      
        </div>
      );
  }
  }
}

const mapStateToProps = ({ User_state }) => {
  return {
      isLoggedIn : User_state.isLoggedIn
  };
};

const mapDispatchToProps = dispatch => ({

    handleLogin: (email, password) => dispatch(loginThunk(email, password)),
    handleRegister: (...args) => dispatch(registerThunk(...args)),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Landing);
