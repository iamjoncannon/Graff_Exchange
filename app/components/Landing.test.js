import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import Landing from './Landing'
import configureStore from 'redux-mock-store';
import thunk from "redux-thunk"
const mockStore = configureStore([thunk]);
import { BrowserRouter as Router } from "react-router-dom";
import { loginThunk, registerThunk } from "../../store/User/thunks_for_User.js";
import nocked from "../../test/nock.setup.js"

const fake_store =  { 
    User_state: {
        balance: 100
    },
    Portfolio_state: {
        portfolio: {
            FB: {
                current_holding: 2,
                price: 100
            }
        },
        selectedPortfolioItem: "FB"
    }
}

const fake_api_data = {"data":{"sign_up":{"email":"c@c.co","first_name":"j","last_name":"c","token":"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6NywiaWF0IjoxNTcwMTU1ODQ0LCJleHAiOjE1NzAyNDIyNDR9.pZ2gy3h-ctD3OjD0Y92TJa253_DG_6ugp4cXuGOtfQg","balance":50000,"holdings":[{"user_data":{"symbol":"GOOGL","current_holding":0,"__typename":"Holding_User_Data"},"ohlc_data":{"latestPrice":1189.43,"companyName":"Alphabet, Inc.","change":11.51,"changePercent":0.00977,"open":1183.95,"__typename":"Holding_OHLC_Data"},"__typename":"Holding"}],"transaction_history":[{"id":"34","type":"Buy","symbol":"GOOGL","quantity":0,"price":0,"date_conducted":"Thu Oct 03 2019 22:24:04 GMT-0400 (Eastern Daylight Time)","__typename":"Transaction"}],"__typename":"User_Profile"}}}

const store_setup = (store) => {

    return renderer.create(
        <Provider store={store} >
           <Router >
               <Landing pathname={"/"} exit={()=>{}}/>
           </Router>
       </Provider>)
}

describe("Landing", ()=>{

    let component 
    let store 
    let connected_react_component

    beforeEach(()=>{

        nocked(fake_api_data)

        window.innerWidth = 1101

        store = mockStore( fake_store );
    
        component = store_setup(store)

        connected_react_component = component.root.children[0].children[0].children[0].children[0]._fiber.stateNode

    })

    it('1) renders successfully with data from redux store', ()=>{

        let tree = component.toJSON();
        
        expect(tree).toMatchSnapshot();
    })

    it("renders 'header' div with 'title' and 'blurb' divs", ()=>{

        const parent = component.root.find(node => node.props.className === "header")

        expect(parent).toBeTruthy()

        expect(parent.children[0].props.className).toEqual("title")
        expect(parent.children[1].props.className).toEqual("blurb")
    })

    it("renders a button with onClick", ()=>{

        const button = component.root.findByType("button")

        expect(button).toBeTruthy()
        expect(button.props.onClick).toBeTruthy()
    })
   
    // if sign up
    it(">> if sign up <<", ()=>{})

    it("renders a span with text: Already have an account?", ()=>{

        renderer.act(()=>{

            connected_react_component.setState({mode: "sign-up"})
        })

        let span = component.root.find(node => node.type === "span" && node.children[0] === ' Already have an account? ') 
    

        expect(span).toBeTruthy()    
    })
   
    it("renders a Sign In span with onClick callback ", ()=>{

        renderer.act(()=>{

            connected_react_component.setState({mode: "sign-up"})
        })

        let span = component.root.find(node => node.type === "span" && node.children[0] === 'Sign In') 

        expect(span).toBeTruthy() 
    })
   
    it("renders a form with five inputs ", ()=>{

        renderer.act(()=>{

            connected_react_component.setState({mode: "sign-up"})
        })

        const form = component.root.findAllByType("form")

        const inputs = component.root.findAllByType("input")

        expect(form.length).toEqual(1)
        expect(inputs.length).toEqual(5)
    })

    it("submit button calls handleRegister", ()=>{

        renderer.act(()=>{

            connected_react_component.setState({mode: "sign-up"})
        })

        const button = component.root.findByType("button")

        const fields = {firstName: 'a', 
                        lastName: "b", 
                        email: "a@b.com", 
                        password: "a", 
                        confirm_password: "a"}

        renderer.act(()=>{
           
            connected_react_component.setState(fields)
           
        })
        
        renderer.act(()=>{

            button.props.onClick()
        })    
        
        store.dispatch( registerThunk(fields) ).then(()=> {

            const dispatched_actions = store.getActions()            
            expect(dispatched_actions.length).toEqual(2)
            expect(dispatched_actions[0].type).toEqual(dispatched_actions[1].type)                
        }) 

    })

})