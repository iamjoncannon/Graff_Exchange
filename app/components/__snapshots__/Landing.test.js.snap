// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Landing 1) renders successfully with data from redux store 1`] = `
<div
  className="landing"
>
  <div
    className="header"
  >
    <span
      className="title"
    >
      Graff Exchange
    </span>
    <span
      className="blurb"
    >
      Graff Exchange, a real-time stock portfolio application built with a full stack: React, Redux, Sass, Socket.io, Node.js, GraphQL, Redis, PostgreSQL, and AWS.
    </span>
    <img
      src="../assets/graphql_logo.png"
    />
    <div
      className="sign-btns"
    >
      <span>
        Don't have an account? 
      </span>
      <span
        onClick={[Function]}
      >
        Sign Up
      </span>
    </div>
  </div>
  <div
    className="form"
  >
    <form>
      <div>
        <span>
          Email
        </span>
        <input
          name="email"
          onChange={[Function]}
          required={true}
          type="email"
          value=""
        />
      </div>
      <div>
        <span>
          Password
        </span>
        <input
          name="password"
          onChange={[Function]}
          type="password"
          value=""
        />
      </div>
      <button
        onClick={[Function]}
      >
        Sign In
      </button>
      <span
        onClick={[Function]}
      >
        Sign into Demo Account
      </span>
    </form>
  </div>
</div>
`;
